README
+=============================================================================+
+                                                                             +
+ Rhythmflow - README (04/01/2010)                                            +
+                                                                             +
+=============================================================================+

Clutter is an open source software library for creating fast, visually
rich and animated graphical user interfaces.

	Clutter currently requires:
		• GLib >= 2.16.0
		• Cairo >= 1.6
		• PangoCairo >= 1.20
		• OpenGL >= 1.2, OpenGL ES 1.1 or OpenGL ES 2.0
		• GLX, SDL, WGL or an EGL Implementation
		• JSON-GLib >= 0.8 (optional)
		• GDK-Pixbuf >= 2.0 (optional)

NEWS
+=============================================================================+
+                                                                             +
+ Rhythmflow - NEWS (04/01/2010)                                              +
+                                                                             +
+=============================================================================+

Rhythmflow 0.1.0 (4/01/2010)
===============================================================================

 * First 'official' release, expect bugs aplenty.
 * Contains:
	o Hopefully enough functionality to build things.
	o Basic gstreamer 0.10 video playback support.
	o Fairly stable API, though in no way totally stable.
	  Expect CHANGES in future versions.
	o Some simple effect.
	o Non complete but hopefully useful API documentation.
 * If you want to help out see the TODO file.

test_embed.py
'''

@author: Helios GILLES
'''
import cluttergtk
import clutter
import gtk
import os

BLACK_BG = clutter.Color(0,0,0,255)
GREEN_BG = clutter.Color(0,255,0,255)

class test(cluttergtk.Embed):
    def __init__(self):
        cluttergtk.Embed.__init__(self)
        self.connect('size-request', self.size_request)
        
    def rf(self):
        self.stage = self.get_stage()
        self.stage.set_color(BLACK_BG) 
        self.stage.connect('key-press-event', self.parseKeyPress)
        
        self.rect = clutter.Rectangle()
        self.rect.set_size(100,100)
        self.rect.set_color(GREEN_BG)
        self.rect.set_reactive(True)
        self.rect.connect("button-press-event", self.left)
        self.stage.add(self.rect)
        
    def parseKeyPress(self,actor,event):
        if event.keyval == clutter.keysyms.q:
            print "cdfgdfgdfgfgdfgdfgdfgdfg"
            
    def left(self,actor,event):
        print "gjhgjhg"
            
    def size_request (self, widget, requisition):
        requisition.width, requisition.height = -1, 250
        self.rect.set_position(500,50)
        

__init__.py
import rb
import rhythmdb
from os import path, listdir
from urllib import url2pathname
import mimetypes
import test_embed
import time
import gobject

IMAGE_NAMES = ['cover', 'album', 'albumart', '.folder', 'folder']

class Rhythmflow(rb.Plugin):
    def __init__ (self):
        rb.Plugin.__init__ (self)
        
    def activate (self, shell):
        self.source = shell.get_property("selected_source")
        self.source.connect("status-changed", self.print_msg)
        self.tmp = 0
            
    def print_msg(self, s):
        self.tmp = self.tmp + 1  
        gobject.timeout_add_seconds(1,self.first_time_load_songs,self.tmp)


    def first_time_load_songs(self,tmp):
        if self.tmp == tmp:
            print "========== Mise a jour de Cover =========="
        return False
  
    def get_cover(self, db_entry=None):
        if db_entry:
            cover_dir = path.dirname(url2pathname(db_entry.get_playback_uri()).replace('file://', ''))
            if path.isdir(cover_dir):
                for f in listdir(cover_dir):
                    file_name = path.join(cover_dir, f)
                    mt = mimetypes.guess_type(file_name)[0]
                    if mt and mt.startswith('image/'):
                        if path.splitext(f)[0].lower() in IMAGE_NAMES:
                            return file_name

            # Find cover saved by artdisplay plugin
            song_info = self.get_song_info(db_entry)
            for rb_cover_path in ('~/.gnome2/rhythmbox/covers', '~/.cache/rhythmbox/covers/'):
                for file_type in ('jpg', 'png', 'jpeg', 'gif', 'svg'):
                    cover_file = path.join(path.expanduser(rb_cover_path),
                                           '%s - %s.%s' %
                                           (song_info['artist'],
                                            song_info['album'],
                                            file_type))
                    if path.isfile(cover_file):
                        return cover_file

            # No cover found
            return None
        # Not playing
        return None
    
    def get_song_info(self, db_entry=None):
        song_info = {}
        db = self.shell.get_property("db")
        if db_entry:
            song_info['title'] = db.entry_get(db_entry, rhythmdb.PROP_TITLE)
            song_info['artist'] = db.entry_get(db_entry, rhythmdb.PROP_ARTIST)
            song_info['album'] = db.entry_get(db_entry, rhythmdb.PROP_ALBUM)
        return song_info

engine/control.py
'''

@author: Helios GILLES
'''

import clutter
import rhythmflow 

DEPTH_CONTROL = 1

PATH = "./img/control/"

class control:
    def __init__(self,parent):
        self.parent = parent
        self.controls = clutter.Group()
        self.play_pause = self.play_pause(self.parent)
        self.stop = self.stop(self.parent)
        self.rewind = self.rewind(self.parent)
        self.forward = self.forward(self.parent)
        self.timeline = clutter.Timeline(30)
        self.timeline.set_duration(200)
        self.alpha = clutter.Alpha(self.timeline,clutter.LINEAR)      
        self.behaviours = []
        
    def play_pause(self,parent):
        play_pause = clutter.Texture()
        play_pause.set_from_file(PATH+"control_play.png")
        play_pause.set_anchor_point(play_pause.get_width()/2,play_pause.get_height()/2)
        play_pause.set_position(play_pause.get_width(),play_pause.get_height())
        play_pause.set_depth(DEPTH_CONTROL)
        self.controls.add(play_pause)
        return play_pause
    
    def stop(self,parent):
        stop = clutter.Texture()
        stop.set_from_file(PATH+"control_stop.png")
        stop.set_anchor_point(stop.get_width()/2,stop.get_height()/2)
        stop.set_position(self.play_pause.get_x()+self.play_pause.get_width()+stop.get_width()/2,self.play_pause.get_y())
        stop.set_depth(DEPTH_CONTROL)
        self.controls.add(stop)
        return stop
    
    def rewind(self,parent):
        rewind = clutter.Texture()
        rewind.set_from_file(PATH+"control_rewind.png")
        rewind.set_anchor_point(rewind.get_width()/2,rewind.get_height()/2)
        rewind.set_position(self.stop.get_x()+self.stop.get_width()+rewind.get_width()/1.5,self.play_pause.get_y())
        rewind.set_depth(DEPTH_CONTROL)
        self.controls.add(rewind)
        return rewind
    
    def forward(self,parent):
        forward = clutter.Texture()
        forward.set_from_file(PATH+"control_forward.png")
        forward.set_anchor_point(forward.get_width()/2,forward.get_height()/2)
        forward.set_position(self.rewind.get_x()+self.rewind.get_width()+forward.get_width()/2,self.play_pause.get_y())
        forward.set_depth(DEPTH_CONTROL)
        self.controls.add(forward)
        return forward
    
if __name__=="__main__":
    rhythmflow.rhythmflow()

engine/scrollbar.py
'''

@author: Helios GILLES
'''

import clutter
import rhythmflow 

DEPTH_SCROLLBAR = 1

PATH = "./img/scrollbar/"

class scrollbar:
    def __init__(self,parent):
        self.parent = parent
        self.decal_lr = 12
        self.scrollbars = clutter.Group()
        self.bg = self.bg(self.parent)
        self.left = self.left(self.parent)
        self.right = self.right(self.parent)
        self.button = self.button(self.parent)
        self.timeline = clutter.Timeline(30)
        self.timeline.set_duration(200)
        self.alpha = clutter.Alpha(self.timeline,clutter.EASE_IN_OUT_SINE)      
        self.behaviours = []
        
    def bg(self,parent):
        bg = clutter.Texture()
        bg.set_from_file(PATH+"scrollbar_bg.png")
        bg.set_width(parent.get_width()/1.8)
        bg.set_anchor_point(bg.get_width()/2,bg.get_height()/2)
        bg.set_position(parent.stage.get_width()/2,parent.stage.get_height()/1.2)
        bg.set_depth(DEPTH_SCROLLBAR)
        self.scrollbars.add(bg)
        return bg 
        
    def left(self,parent):
        left = clutter.Texture()
        left.set_from_file(PATH+"scrollbar_left.png")
        left.set_anchor_point(left.get_width()/2,left.get_height()/2)
        left.set_position(self.bg.get_x()-self.bg.get_width()/2-left.get_width()/2+self.decal_lr,self.bg.get_y())
        left.set_depth(DEPTH_SCROLLBAR)
        left.set_reactive(True)
        left.connect("button-press-event", parent.click_scrollbar_left)
        self.scrollbars.add(left)
        return left
    
    def right(self,parent):
        right = clutter.Texture()
        right.set_from_file(PATH+"scrollbar_right.png")
        right.set_anchor_point(right.get_width()/2,right.get_height()/2)
        right.set_position(self.bg.get_x()+self.bg.get_width()/2+right.get_width()/2-self.decal_lr,self.bg.get_y())
        right.set_depth(DEPTH_SCROLLBAR)
        right.set_reactive(True)
        right.connect("button-press-event", parent.click_scrollbar_right)
        self.scrollbars.add(right)
        return right
    
    def button(self,parent):
        button = clutter.Texture()
        button.set_from_file(PATH+"scrollbar_button.png")
        button.set_anchor_point(button.get_width()/2,button.get_height()/2)
        button.set_position(self.bg.get_x()-self.bg.get_width()/2+button.get_width()/2,self.bg.get_y())
        button.set_depth(DEPTH_SCROLLBAR)
        self.scrollbars.add(button)
        return button
    
    def init_scrollbar(self,parent,index_cover_selected,size_covers):
        if index_cover_selected == 0:
            self.button.set_position(self.bg.get_x()-self.bg.get_width()/2+self.button.get_width()/2,self.bg.get_y())
        elif index_cover_selected == size_covers - 1:
            self.button.set_position(self.bg.get_x()+self.bg.get_width()/2-self.button.get_width()/2,self.bg.get_y())
        else:
            self.button.set_position(self.bg.get_x()-self.bg.get_width()/2+self.button.get_width()/2+index_cover_selected*(self.bg.get_width()-self.button.get_width())/(size_covers-1),self.bg.get_y())
    
    def anim_left(self,parent,index_cover_selected,size_covers):
        self.behaviours = []
        if index_cover_selected == 0:
            path = clutter.Path("M "+str(self.button.get_x())+" "+str(self.button.get_y())+" L "+str(self.bg.get_x()-self.bg.get_width()/2+self.button.get_width()/2)+" "+str(self.bg.get_y()))
        elif index_cover_selected == size_covers - 1:
            path = clutter.Path("M "+str(self.button.get_x())+" "+str(self.button.get_y())+" L "+str(self.bg.get_x()+self.bg.get_width()/2-self.button.get_width()/2)+" "+str(self.bg.get_y()))
        else:
            path = clutter.Path("M "+str(self.button.get_x())+" "+str(self.button.get_y())+" L "+str(self.bg.get_x()-self.bg.get_width()/2+self.button.get_width()/2+index_cover_selected*(self.bg.get_width()-self.button.get_width())/(size_covers-1))+" "+str(self.bg.get_y()))
        behavior = clutter.BehaviourPath(self.alpha,path)
        behavior.apply(self.button)
        self.behaviours.append(behavior)  
        self.timeline.start()
        
    def anim_right(self,parent,index_cover_selected,size_covers):
        self.behaviours = []
        if index_cover_selected == 0:
            path = clutter.Path("M "+str(self.button.get_x())+" "+str(self.button.get_y())+" L "+str(self.bg.get_x()-self.bg.get_width()/2+self.button.get_width()/2)+" "+str(self.bg.get_y()))
        elif index_cover_selected == size_covers - 1:
            path = clutter.Path("M "+str(self.button.get_x())+" "+str(self.button.get_y())+" L "+str(self.bg.get_x()+self.bg.get_width()/2-self.button.get_width()/2)+" "+str(self.bg.get_y()))
        else:
            path = clutter.Path("M "+str(self.button.get_x())+" "+str(self.button.get_y())+" L "+str(self.bg.get_x()-self.bg.get_width()/2+self.button.get_width()/2+index_cover_selected*(self.bg.get_width()-self.button.get_width())/(size_covers-1))+" "+str(self.bg.get_y()))
        behavior = clutter.BehaviourPath(self.alpha,path)
        behavior.apply(self.button)
        self.behaviours.append(behavior)  
        self.timeline.start()         
        
if __name__=="__main__":
    rhythmflow.rhythmflow()

engine/fx_default.py
'''

@author: Helios GILLES
'''

import clutter
import rhythmflow 

ANGLE = 45
BIG_SPACE = 0.1
SPACE = 0.4
SCALE = 1.2

class fx:
    def __init__(self,parent):
        self.parent = parent
        self.timeline = clutter.Timeline(30)
        self.timeline.set_duration(200)
        self.alpha = clutter.Alpha(self.timeline,clutter.LINEAR)   
        self.view = clutter.Group()   
        self.behaviours = []
        self.anim_finished = True
        self.timeline.connect("started", self.event_anim_started)
        self.timeline.connect("completed", self.event_anim_completed)
        
    def event_anim_started(self,event):
        self.anim_finished = False
    
    def event_anim_completed(self,event):
        self.anim_finished = True
                
    def get_view(self):
        return self.view
    
    def init_fx(self,parent,covers,index_cover_selected):
        for (index, cover) in enumerate(covers):
            self.view.add(cover)
            cover.set_reactive(True)
            cover.connect("button-press-event", parent.click_cover)
            if index == index_cover_selected:
                cover.set_scale(SCALE,SCALE)
                cover.set_position(parent.get_width()/2,parent.get_height()/2)
                cover.set_depth(0)
            elif index < index_cover_selected:
                cover.set_rotation(clutter.Y_AXIS,ANGLE,0,0,0)
                cover.set_position(parent.get_width()/2-cover.get_width()*BIG_SPACE+cover.get_width()*(index-index_cover_selected)*SPACE,parent.get_height()/2)
                cover.set_depth(index-index_cover_selected)
            else:
                cover.set_rotation(clutter.Y_AXIS,-ANGLE,0,0,0)
                cover.set_position(parent.get_width()/2+cover.get_width()*BIG_SPACE+cover.get_width()*(index-index_cover_selected)*SPACE,parent.get_height()/2)
                cover.set_depth(index_cover_selected-index)
        
    def anim_left(self,parent,covers,index_cover_selected):
        print "left"
        if self.anim_finished :
            if (parent.get_index_cover_selected() > 0):
                self.behaviours = []
                for (index, cover) in enumerate(covers):
                    if index < index_cover_selected - 1:
                        cover.set_depth(index-index_cover_selected)
                        path = clutter.Path("M "+str(cover.get_x())+" "+str(cover.get_y())+" L "+str(cover.get_x()+cover.get_width()*SPACE)+" "+str(cover.get_y()))
                        behavior = clutter.BehaviourPath(self.alpha,path)
                        behavior.apply(cover)
                        self.behaviours.append(behavior)
                
                covers[index_cover_selected-1].set_depth(0)      
                behavior = clutter.BehaviourRotate(clutter.Y_AXIS,ANGLE,0,self.alpha,clutter.ROTATE_CCW)
                behavior.apply(covers[index_cover_selected-1])
                self.behaviours.append(behavior)
                (cur_scale_x,cur_scale_y) = covers[index_cover_selected-1].get_scale()
                behavior = clutter.BehaviourScale(cur_scale_x,cur_scale_y,1.2,1.2,self.alpha)
                behavior.apply(covers[index_cover_selected-1])
                self.behaviours.append(behavior)
                path = clutter.Path("M "+str(covers[index_cover_selected-1].get_x())+" "+str(covers[index_cover_selected-1].get_y())+" L "+str(parent.get_width()/2)+" "+str(parent.get_height()/2))
                behavior = clutter.BehaviourPath(self.alpha,path)
                behavior.apply(covers[index_cover_selected-1])
                self.behaviours.append(behavior)
             
                covers[index_cover_selected].set_depth(-1)
                behavior = clutter.BehaviourRotate(clutter.Y_AXIS,0,360-ANGLE,self.alpha,clutter.ROTATE_CCW)
                behavior.apply(covers[index_cover_selected])
                self.behaviours.append(behavior)
                (cur_scale_x,cur_scale_y) = covers[index_cover_selected].get_scale()
                behavior = clutter.BehaviourScale(cur_scale_x,cur_scale_y,1,1,self.alpha)
                behavior.apply(covers[index_cover_selected])
                self.behaviours.append(behavior)
                path = clutter.Path("M "+str(covers[index_cover_selected].get_x())+" "+str(covers[index_cover_selected].get_y())+" L "+str(parent.get_width()/2+covers[index_cover_selected].get_width()*BIG_SPACE+covers[index_cover_selected].get_width()*(1)*SPACE)+" "+str(covers[index_cover_selected].get_y()))
                behavior = clutter.BehaviourPath(self.alpha,path)
                behavior.apply(covers[index_cover_selected])
                self.behaviours.append(behavior)
                
                for (index, cover) in enumerate(covers):
                    if index > index_cover_selected:
                        cover.set_depth(index_cover_selected-index-2)
                        path = clutter.Path("M "+str(cover.get_x())+" "+str(cover.get_y())+" L "+str(cover.get_x()+cover.get_width()*SPACE)+" "+str(cover.get_y()))
                        behavior = clutter.BehaviourPath(self.alpha,path)
                        behavior.apply(cover)
                        self.behaviours.append(behavior)
                        
                self.behaviours.append(behavior)        
                self.timeline.start() 
                parent.dec_index_cover_selected()

            
    def anim_right(self,parent,covers,index_cover_selected):
        print "rigth"
        if self.anim_finished :
            if (parent.get_index_cover_selected() < len(covers)-1):
                self.behaviours = []
                for (index, cover) in enumerate(covers):
                    if index < index_cover_selected:
                        cover.set_depth(index-index_cover_selected)
                        path = clutter.Path("M "+str(cover.get_x())+" "+str(cover.get_y())+" L "+str(cover.get_x()-cover.get_width()*SPACE)+" "+str(cover.get_y()))
                        behavior = clutter.BehaviourPath(self.alpha,path)
                        behavior.apply(cover)
                        self.behaviours.append(behavior)
                        
                covers[index_cover_selected].set_depth(-1) 
                behavior = clutter.BehaviourRotate(clutter.Y_AXIS,0,ANGLE,self.alpha,clutter.ROTATE_CW)
                behavior.apply(covers[index_cover_selected])
                self.behaviours.append(behavior)
                (cur_scale_x,cur_scale_y) = covers[index_cover_selected].get_scale()
                behavior = clutter.BehaviourScale(cur_scale_x,cur_scale_y,1,1,self.alpha)
                behavior.apply(covers[index_cover_selected])
                self.behaviours.append(behavior)
                path = clutter.Path("M "+str(covers[index_cover_selected].get_x())+" "+str(covers[index_cover_selected].get_y())+" L "+str(parent.get_width()/2-covers[index_cover_selected].get_width()*BIG_SPACE-covers[index_cover_selected].get_width()*(1)*SPACE)+" "+str(covers[index_cover_selected].get_y()))
                behavior = clutter.BehaviourPath(self.alpha,path)
                behavior.apply(covers[index_cover_selected])
                self.behaviours.append(behavior)
     
                covers[index_cover_selected+1].set_depth(0) 
                behavior = clutter.BehaviourRotate(clutter.Y_AXIS,ANGLE,0,self.alpha,clutter.ROTATE_CCW)
                behavior.apply(covers[index_cover_selected+1])
                self.behaviours.append(behavior)
                (cur_scale_x,cur_scale_y) = covers[index_cover_selected+1].get_scale()
                behavior = clutter.BehaviourScale(cur_scale_x,cur_scale_y,1.2,1.2,self.alpha)
                behavior.apply(covers[index_cover_selected+1])
                self.behaviours.append(behavior)
                path = clutter.Path("M "+str(covers[index_cover_selected+1].get_x())+" "+str(covers[index_cover_selected+1].get_y())+" L "+str(parent.get_width()/2)+" "+str(parent.get_height()/2))
                behavior = clutter.BehaviourPath(self.alpha,path)
                behavior.apply(covers[index_cover_selected+1])
                self.behaviours.append(behavior) 
                
                for (index, cover) in enumerate(covers):
                    if index > index_cover_selected + 1:
                        cover.set_depth(index_cover_selected-index-1)
                        path = clutter.Path("M "+str(cover.get_x())+" "+str(cover.get_y())+" L "+str(cover.get_x()-cover.get_width()*SPACE)+" "+str(cover.get_y()))
                        behavior = clutter.BehaviourPath(self.alpha,path)
                        behavior.apply(cover)
                        self.behaviours.append(behavior)
                        
                self.behaviours.append(behavior)        
                self.timeline.start() 
                parent.inc_index_cover_selected()
                
    def anim_dbl_click(self,actor):
        print "dbl_click"
        self.behaviours = []
        path = clutter.Path("M "+str(actor.get_x())+" "+str(actor.get_y())+" L "+str(actor.get_x())+" "+str(actor.get_y()-100)+" L "+str(actor.get_x())+" "+str(actor.get_y()))
        behavior = clutter.BehaviourPath(self.alpha,path)
        behavior.apply(actor)
        self.behaviours.append(behavior)
        self.timeline.start() 

if __name__=="__main__":
    rhythmflow.rhythmflow()
        

engine/fx_setanta.py
'''

@author: Helios GILLES
'''

import clutter
import rhythmflow 

ANGLE = 45
BIG_SPACE = 0.88
SPACE = 0.36
SCALE_CENTER = 1.12
SCALE_OTHER = 0.91

class fx:
    def __init__(self,parent):
        self.parent = parent
        self.timeline = clutter.Timeline(30)
        self.timeline.set_duration(200)
        self.alpha = clutter.Alpha(self.timeline,clutter.LINEAR)   
        self.view = clutter.Group()   
        self.behaviours = []
        self.anim_finished = True
        self.timeline.connect("started", self.event_anim_started)
        self.timeline.connect("completed", self.event_anim_completed)
        
    def event_anim_started(self,event):
        self.anim_finished = False
    
    def event_anim_completed(self,event):
        self.anim_finished = True
                
    def get_view(self):
        return self.view
    
    def init_fx(self,parent,covers,index_cover_selected):
        for (index, cover) in enumerate(covers):
            self.view.add(cover)
            cover.set_reactive(True)
            cover.connect("button-press-event", parent.click_cover)
            if index == index_cover_selected:
                cover.set_scale(SCALE_CENTER,SCALE_CENTER)
                cover.set_position(parent.get_width()/2,parent.get_height()/2)
                cover.set_depth(0)    
            elif index == index_cover_selected - 1:
                cover.set_rotation(clutter.Y_AXIS,-ANGLE,0,0,0)
                cover.set_position(parent.get_width()/2-cover.get_width()*BIG_SPACE,parent.get_height()/2)
                cover.set_depth(-1)
            elif index == index_cover_selected + 1:
                cover.set_rotation(clutter.Y_AXIS,ANGLE,0,0,0)
                cover.set_position(parent.get_width()/2+cover.get_width()*BIG_SPACE,parent.get_height()/2)
                cover.set_depth(-1)
            elif index < index_cover_selected - 1:
                cover.set_scale(SCALE_OTHER,SCALE_OTHER)
                cover.set_position(parent.get_width()/2-cover.get_width()*BIG_SPACE+cover.get_width()*SPACE*(index-index_cover_selected),parent.get_height()/2)
                cover.set_depth(index-index_cover_selected)
            else:
                cover.set_scale(SCALE_OTHER,SCALE_OTHER)
                cover.set_position(parent.get_width()/2+cover.get_width()*BIG_SPACE+cover.get_width()*SPACE*(index-index_cover_selected),parent.get_height()/2)
                cover.set_depth(index_cover_selected-index)
        
    def anim_left(self,parent,covers,index_cover_selected):
        print "left"
        if self.anim_finished :
            if (parent.get_index_cover_selected() > 0):
                self.behaviours = []
                for (index, cover) in enumerate(covers):
                    if index < index_cover_selected - 1:
                        cover.set_depth(index-index_cover_selected)
                        path = clutter.Path("M "+str(cover.get_x())+" "+str(cover.get_y())+" L "+str(cover.get_x()+cover.get_width()*SPACE)+" "+str(cover.get_y()))
                        behavior = clutter.BehaviourPath(self.alpha,path)
                        behavior.apply(cover)
                        self.behaviours.append(behavior)
                
                covers[index_cover_selected-1].set_depth(0)      
                behavior = clutter.BehaviourRotate(clutter.Y_AXIS,ANGLE,0,self.alpha,clutter.ROTATE_CCW)
                behavior.apply(covers[index_cover_selected-1])
                self.behaviours.append(behavior)
                (cur_scale_x,cur_scale_y) = covers[index_cover_selected-1].get_scale()
                behavior = clutter.BehaviourScale(cur_scale_x,cur_scale_y,1.2,1.2,self.alpha)
                behavior.apply(covers[index_cover_selected-1])
                self.behaviours.append(behavior)
                path = clutter.Path("M "+str(covers[index_cover_selected-1].get_x())+" "+str(covers[index_cover_selected-1].get_y())+" L "+str(parent.get_width()/2)+" "+str(parent.get_height()/2))
                behavior = clutter.BehaviourPath(self.alpha,path)
                behavior.apply(covers[index_cover_selected-1])
                self.behaviours.append(behavior)
             
                covers[index_cover_selected].set_depth(-1)
                behavior = clutter.BehaviourRotate(clutter.Y_AXIS,0,360-ANGLE,self.alpha,clutter.ROTATE_CCW)
                behavior.apply(covers[index_cover_selected])
                self.behaviours.append(behavior)
                (cur_scale_x,cur_scale_y) = covers[index_cover_selected].get_scale()
                behavior = clutter.BehaviourScale(cur_scale_x,cur_scale_y,1,1,self.alpha)
                behavior.apply(covers[index_cover_selected])
                self.behaviours.append(behavior)
                path = clutter.Path("M "+str(covers[index_cover_selected].get_x())+" "+str(covers[index_cover_selected].get_y())+" L "+str(parent.get_width()/2+covers[index_cover_selected].get_width()*BIG_SPACE+covers[index_cover_selected].get_width()*(1)*SPACE)+" "+str(covers[index_cover_selected].get_y()))
                behavior = clutter.BehaviourPath(self.alpha,path)
                behavior.apply(covers[index_cover_selected])
                self.behaviours.append(behavior)
                
                for (index, cover) in enumerate(covers):
                    if index > index_cover_selected:
                        cover.set_depth(index_cover_selected-index-2)
                        path = clutter.Path("M "+str(cover.get_x())+" "+str(cover.get_y())+" L "+str(cover.get_x()+cover.get_width()*SPACE)+" "+str(cover.get_y()))
                        behavior = clutter.BehaviourPath(self.alpha,path)
                        behavior.apply(cover)
                        self.behaviours.append(behavior)
                        
                self.behaviours.append(behavior)        
                self.timeline.start() 
                parent.dec_index_cover_selected()

            
    def anim_right(self,parent,covers,index_cover_selected):
        print "rigth"
        if self.anim_finished :
            if (parent.get_index_cover_selected() < len(covers)-1):
                self.behaviours = []
                for (index, cover) in enumerate(covers):
                    if index < index_cover_selected:
                        cover.set_depth(index-index_cover_selected)
                        path = clutter.Path("M "+str(cover.get_x())+" "+str(cover.get_y())+" L "+str(cover.get_x()-cover.get_width()*SPACE)+" "+str(cover.get_y()))
                        behavior = clutter.BehaviourPath(self.alpha,path)
                        behavior.apply(cover)
                        self.behaviours.append(behavior)
                        
                covers[index_cover_selected].set_depth(-1) 
                behavior = clutter.BehaviourRotate(clutter.Y_AXIS,0,ANGLE,self.alpha,clutter.ROTATE_CW)
                behavior.apply(covers[index_cover_selected])
                self.behaviours.append(behavior)
                (cur_scale_x,cur_scale_y) = covers[index_cover_selected].get_scale()
                behavior = clutter.BehaviourScale(cur_scale_x,cur_scale_y,1,1,self.alpha)
                behavior.apply(covers[index_cover_selected])
                self.behaviours.append(behavior)
                path = clutter.Path("M "+str(covers[index_cover_selected].get_x())+" "+str(covers[index_cover_selected].get_y())+" L "+str(parent.get_width()/2-covers[index_cover_selected].get_width()*BIG_SPACE-covers[index_cover_selected].get_width()*(1)*SPACE)+" "+str(covers[index_cover_selected].get_y()))
                behavior = clutter.BehaviourPath(self.alpha,path)
                behavior.apply(covers[index_cover_selected])
                self.behaviours.append(behavior)
     
                covers[index_cover_selected+1].set_depth(0) 
                behavior = clutter.BehaviourRotate(clutter.Y_AXIS,ANGLE,0,self.alpha,clutter.ROTATE_CCW)
                behavior.apply(covers[index_cover_selected+1])
                self.behaviours.append(behavior)
                (cur_scale_x,cur_scale_y) = covers[index_cover_selected+1].get_scale()
                behavior = clutter.BehaviourScale(cur_scale_x,cur_scale_y,1.2,1.2,self.alpha)
                behavior.apply(covers[index_cover_selected+1])
                self.behaviours.append(behavior)
                path = clutter.Path("M "+str(covers[index_cover_selected+1].get_x())+" "+str(covers[index_cover_selected+1].get_y())+" L "+str(parent.get_width()/2)+" "+str(parent.get_height()/2))
                behavior = clutter.BehaviourPath(self.alpha,path)
                behavior.apply(covers[index_cover_selected+1])
                self.behaviours.append(behavior) 
                
                for (index, cover) in enumerate(covers):
                    if index > index_cover_selected + 1:
                        cover.set_depth(index_cover_selected-index-1)
                        path = clutter.Path("M "+str(cover.get_x())+" "+str(cover.get_y())+" L "+str(cover.get_x()-cover.get_width()*SPACE)+" "+str(cover.get_y()))
                        behavior = clutter.BehaviourPath(self.alpha,path)
                        behavior.apply(cover)
                        self.behaviours.append(behavior)
                        
                self.behaviours.append(behavior)        
                self.timeline.start() 
                parent.inc_index_cover_selected()

if __name__=="__main__":
    rhythmflow.rhythmflow()
        

engine/rhythmflow.py
'''

@author: Helios GILLES
'''
import cluttergtk
import clutter
import gtk
import os
import cover
import fx_default
import scrollbar
import control

WIDTH = 1000
HEIGHT = 600
BLACK_BG = clutter.Color(0,0,0,255)

class rhythmflow(cluttergtk.Embed):
    def __init__(self):
        cluttergtk.Embed.__init__(self)
        
    def rf(self):
        self.stage = self.get_stage()
        self.stage.set_size(WIDTH,HEIGHT)
        self.stage.set_color(BLACK_BG) 
        self.stage.connect('key-press-event', self.parseKeyPress)
        self.index_cover_selected = 2
        self.cover = cover.cover(self)
        self.fx = fx_default.fx(self)
        self.fx.init_fx(self, self.get_covers(), self.get_index_cover_selected())
        self.scrollbar = scrollbar.scrollbar(self)
        self.scrollbar.init_scrollbar(self, self.get_index_cover_selected(), self.get_size_covers())
        self.control = control.control(self)
        self.show()
        
    def parseKeyPress(self,actor,event):
        if event.keyval == clutter.keysyms.Left:
            self.anim_left()
        if event.keyval == clutter.keysyms.Right:
            self.anim_right()
        if event.keyval == clutter.keysyms.q:
            exit(0)
            
    def click_scrollbar_left(self,actor,event):
        self.anim_left()
        
    def click_scrollbar_right(self,actor,event):
        self.anim_right()
        
    def click_cover(self,actor,event):
        self.anim_dbl_click(actor)
            
    def anim_left(self):
        self.fx.anim_left(self,self.get_covers(),self.get_index_cover_selected())
        self.scrollbar.anim_left(self,self.get_index_cover_selected(),self.get_size_covers())
        
    def anim_right(self):
        self.fx.anim_right(self,self.get_covers(),self.get_index_cover_selected())
        self.scrollbar.anim_right(self,self.get_index_cover_selected(),self.get_size_covers())    
            
    def anim_dbl_click(self,actor):
        self.fx.anim_dbl_click(actor)
    
    def show(self):
        self.stage.remove_all()
        self.stage.add(self.scrollbar.scrollbars)
        self.stage.add(self.control.controls)
        self.stage.add(self.fx.get_view())
        self.stage.show_all()
        
    def get_width(self):
        return self.stage.get_width()
    
    def get_height(self):
        return self.stage.get_height()
    
    def get_index_cover_selected(self):
        return self.index_cover_selected
    
    def inc_index_cover_selected(self):
        self.index_cover_selected = self.index_cover_selected + 1
    
    def dec_index_cover_selected(self):
        self.index_cover_selected = self.index_cover_selected - 1
        
    def get_size_covers(self):
        return len(self.cover.covers)
    
    def get_covers(self):
        return self.cover.covers
    
    def set_covers(self,covers):
        self.cover.covers = covers
    
    def get_list_covers(self):
        return self.cover.list_covers
    
    def set_list_covers(self,list_covers):
        self.cover.list_covers = list_covers
        
    def get_scrollbars(self):
        return self.scrollbar.scrollbars
    
    def set_scrollbars(self,scrollbars):
        self.scrollbar.scrollbars = scrollbars
    
    def get_controls(self):
        return self.control.controls
    
    def set_controls(self,controls):
        self.control.controls = controls
    
    def get_fx(self):
        return self.fx
    
if __name__=="__main__":
    window = gtk.Window()
    window.connect('destroy', gtk.main_quit)
    window.set_title('Rhythmflow 0.1')
    vbox = gtk.VBox(False, 1)
    window.add(vbox)
    rf = rhythmflow()
    vbox.pack_start(rf, True, True, 0)
    rf.set_size_request(1000, 600)
    rf.realize()
    rf.rf()
    window.set_position(gtk.WIN_POS_CENTER)
    window.show_all()

    gtk.main()